lin_guider/doc/lg_tool.man
.\" -*- nroff -*-
.TH lg_tool 1 "November 2016" "lg_tool(1)" "lg_tool manual page"
.SH NAME
lg_tool - command tool to send commands to lin_guider.
.SH SYNOPSIS
.br
lg_tool get_ver [-v]
.br
lg_tool guider [-v] start|stop
.br
lg_tool status [-v]
.br
lg_tool dither [-v]
.br
lg_tool dither_no_wait [-v] rX rY
.br
lg_tool get_distance [-v]
.br
lg_tool get_ra_dec_drift [-v]
.br
lg_tool set_dithering_range [-v] range
.br
lg_tool set_square_pos [-v] X Y
.br
lg_tool set_ovls_pos [-v] X Y
.br
lg_tool set_reticle_pos [-v] X Y
.br
lg_tool find_star [-v]
.br
lg_tool save_frame [-v] filename
.br
lg_tool save_frame_decorated [-v] filename

.SH DESCRIPTION
lg_tool is used to control lin_guider. It is particularly useful as dithering agent executed between exposures by the imaging software to perform dithering.

.SH OPTIONS
.br
.B -v
verbose output
.br
.B -h
print this help

.SH VARIABLES
.B LG_SOCKET
environment variable can be set to change the default.
The default is '/tmp/lg_ss'. To use TCP connection use
format 'tcp://hostname:port'.

.SH SEE ALSO
.B lin_guider(7)

.SH COPYRIGHT AND LICENSE

Copyright (C) 2008-2016 Rumen G.Bogdanovski

This is a free software, you can redistribute it and/or modify
it under the terms of GPL3

The authors assume no liability or responsibility for damage or injury
to persons or property arising from any use of this product. Use it at
your own risk.

.SH BUGS
If you find any, please send bug reports to rumen@skyarchive.org


lin_guider/ChangeLog
2016-12-16 Lin_guider 4.1.1 (c)Galaxy Master <galaxy_master@mail.ru>

	* Add manpages for lg_tool and lin_guider, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Minor fixes, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2016-10-04 Lin_guider 4.1.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Guider start/stop and status remote commands added
	* Fixes for modern g++ v.6+ (with backward compatibility)
	* Added SET_GUIDER_OVLS_POS, SET_GUIDER_RETICLE_POS, SET_DITHERING_RANGE, FIND_STAR and
	  GET_RA_DEC_DRIFT commands, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Unify messages of the command server, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* lg_tool.pl now supports the new commands, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use doubleclick to move the overlays (square and guide frame), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use Ctrl + doubleclick to move the reticle, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Drift graph Y scale can now show 1"/cell, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* ASI driver: Fixed to work with the latest (v.0.3.0908) and previous ASI SDKs, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* ASI driver: fix segmentation fault if symbol is not available in libasucamera2.so, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest Atik SDK (v.1.5), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Atik driver: fix segmentation fault if symbol is not available in libatikccd.so, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2016-07-14 Lin_guider 4.0.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Created infrastructure to support selectable drift graphs
	* Created infrastructure to support selectable guiding algorithms
	* Changing Graph X and Y cells does not require restart any more
	* Added guiding status line to display guiding messages
	* GUI Tweaks
	* Added two versions Target drift graph, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Enhanced Scrolling drift graph, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Added Enhanced DONUTS guiding algorithm, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Gain of Null camera controls the simuated frame noise, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Minor fixes in ASI camera driver, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest ASI SDK (v.0.3.0623), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Overlays (square and guide frame) are dragged relative to the mouse position not the center, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2016-04-25 Lin_guider 3.3.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Add pulse driver for SkyWatcher/Orion mounts (HC firmware at least 3.38.08 or 4.38.08 is required), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add colour support for Atik Cameras, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest Atik SDK (v.1.1), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest ASI SDK (v.0.3.0324), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* UI tweaks, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2016-03-18 Lin_guider 3.2.2 (c)Galaxy Master <galaxy_master@mail.ru>

	* Use latest ASI SDK (v.0.3.0316), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add SAVE_FRAME_DECORATED server command, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2015-11-25 Lin_guider 3.2.1 (c)Galaxy Master <galaxy_master@mail.ru>

	* Use latest Atik SDK (0.41), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest ASI SDK (0.2.1026), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Fix gain control overflow in ASI driver, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add save_frame in lg_tool.pl, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2015-11-03 Lin_guider 3.2.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Fixed DEC swapping in GPIO driver
	* Guider window UI tweaks
	* Add adjust window to fit image button
	* Add support for all Bayer patterns, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Always use '.' in floating point numbers, regardless of the locale, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add normalization of gain coefficients, if enabled changing the guide rate will affect
	  directly the guiding, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Restore reticle angle after restart, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest Atik SDK (0.34), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Use latest ASI SDK (0.1.0803), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add colour support for ASI cameras, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add 16-bit support for ASI cameras, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Fix seeveral issues in ASI driver, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Small fix in Starlight Xpress driver, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2015-04-12 Lin_guider 3.1.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Make reticle and square dragging responsive (independent of the exposure time)
	* Reticle can be moved if the square is over it, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Last used square is now saved in configuration, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Make ASI buffer cleaning optional for better performance, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Add new dithering commands GET_DISTANCE and DITHER_NO_WAIT_XY, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Accept commands over TCP or Unix Sockets(default), by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Make Atik and SX drivers prefer the first camera with ST4 port, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Fixes in Atik, ASI and QHY5 drivers, by Rumen G.Bogdanovski <rumen@skyarchive.org>
	* Created lg_tool.pl to be used as a dithering agent or as standalone application, by Rumen G.Bogdanovski <rumen@skyarchive.org>

2015-02-26 Lin_guider 3.0.0 (c)Galaxy Master <galaxy_master@mail.ru>

	* Added support for ZWO ASI cameras, by Rumen G.Bogdanovski <rumen@skyarchive.org>

lin_guider/tools/lg_tool.pl
#!/usr/bin/perl
# lg_tool.pl
#
#      Author: Rumen G. Bogdanovski <rumen@skyarchive.org>
#
# This file is part of Lin_guider.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use Getopt::Std;
use IO::Socket;
use File::Basename;

my $VERSION = "0.4";
my $verbose = 0;

my %command_val = (
	'GET_VER' => 1,
	'SET_GUIDER_SQUARE_POS' => 2,
	'SAVE_FRAME' => 3,
	'DITHER' => 4,
	'DITHER_NO_WAIT_XY' => 5,
	'GET_DISTANCE' => 6,
	'SAVE_FRAME_DECORATED' => 7,
	'GUIDER' => 8,
	'GET_GUIDER_STATE' => 9,
	'SET_GUIDER_OVLS_POS' => 10,
	'SET_GUIDER_RETICLE_POS' => 11,
	'FIND_STAR' => 12,
	'SET_DITHERING_RANGE' => 13,
	'GET_RA_DEC_DRIFT' => 14
);
my %command_name = reverse %command_val;

use constant HDR_LEN => 8;
#use constant SOCK_NAME => 'tcp://127.0.0.1:5656';
use constant SOCK_NAME => '/tmp/lg_ss';
my $sock_name;


sub print_help() {
	my $N = basename($0);
	print "\n".
	      "Lin-guider command tool v.$VERSION.\n".
	      "This is a GPL software, created by Rumen G.Bogdanovski.\n".
	      "\n".
	      "Usage: $N get_ver [-v]\n".
	      "       $N guider [-v] start|stop\n".
	      "       $N status [-v]\n".
	      "       $N dither [-v]\n".
	      "       $N dither_no_wait [-v] rX rY\n".
	      "       $N get_distance [-v]\n".
	      "       $N get_ra_dec_drift [-v]\n".
	      "       $N set_dithering_range [-v] rage\n".
	      "       $N set_square_pos [-v] X Y\n".
	      "       $N set_ovls_pos [-v] X Y\n".
	      "       $N set_reticle_pos [-v] X Y\n".
	      "       $N find_star [-v]\n".
	      "       $N save_frame [-v] filename\n".
	      "       $N save_frame_decorated [-v] filename\n".
	      "options:\n".
	      "       -v verbose output\n".
	      "       -h print this help\n\n".
	      "       LG_SOCKET environment variable can be set to change the default.\n".
	      "       The default is '".SOCK_NAME."'. To use TCP connection use\n".
	      "       format 'tcp://hostname:port'.\n\n";
}


#
# Communication routines
#
sub lg_connect($) {
	my ($addr) = @_;

	my $sock;
	if ($addr =~ /^tcp:\/\//) {
		$addr =~ s/^tcp:\/\///;
		$sock = IO::Socket::INET->new($addr) or return undef;
	} else {
		$addr =~ s/^unix:\/\///; # Acept "unix://" prefix, not mandatory
		$sock = IO::Socket::UNIX->new($addr) or return undef;
	}
	return $sock;
}


sub lg_disconnect($) {
	my ($sock) = @_;
	close($sock);
}


sub lg_send_cmd($$$) {
	my ($sock, $cmd, $param) = @_;

	my $len = length $param;

	my $cmd = pack('SSL', 2, $cmd, $len);
	if(HDR_LEN != $sock->send($cmd)) {
		return undef;
	}

	if (($len > 0) and ($len != $sock->send($param))) {
		return undef;
	}

	return 1;
}


sub lg_recv_resp($) {
	my ($sock) = @_;

	my $resp;

	$sock->recv($resp, HDR_LEN);
	if (HDR_LEN != length($resp)) {
		return undef;
	}

	my($wrp, $cmd, $len) = unpack ('SSL', $resp);
	if ($len > 0) {
		$sock->recv($resp, $len);
		if ($len != length($resp)) {
			return undef;
		}
        }

	return ($resp, $cmd);
}


sub lg_chat($$) {
	my ($cmd, $paramstr) = @_;

	my $resp;
	my $timeout = 60;  # Timeout for send command and receive relpy

	my $sock = lg_connect($sock_name) or die "Connect to $sock_name: $!\n";
	eval {
		local $SIG{ALRM} = sub { die "Error: can not set timeout\n" };
		alarm($timeout);
		if (! lg_send_cmd($sock, $cmd ,$paramstr)) {
			print STDERR "Send command error.\n";
			lg_disconnect($sock);
			alarm(0);
			return undef;
		}
		($resp, $cmd) = lg_recv_resp($sock);
		lg_disconnect($sock);
		alarm(0);
	};
	if ($@) {
		return ("Error: Operation Timed Out", $cmd);
	}
	return ($resp, $cmd);
}


#
# Lin-guider Commands
#

sub guider {
	my @params = @_;
	if ($#params != 0) {
		print STDERR "guider: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{GUIDER}, $params[0]);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub status {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "status: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{GET_GUIDER_STATE} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub dither {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "dither: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{DITHER} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub get_ver {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "get_ver: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{GET_VER} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^v./) {
		return 1;
	} else {
		return undef;
	}
}


sub get_distance {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "get_distance: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{GET_DISTANCE} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^Error/) {
		return undef;
	} else {
		return 1;
	}
}


sub get_ra_dec_drift {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "get_ra_dec_drift: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{GET_RA_DEC_DRIFT} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^Error/) {
		return undef;
	} else {
		return 1;
	}
}


sub dither_no_wait {
	my @params = @_;
	if ($#params != 1) {
		print STDERR "dither_no_wait: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%.2f %.2f", $params[0], $params[1]);
	my ($resp,$cmd) = lg_chat($command_val{DITHER_NO_WAIT_XY}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub set_square_pos {
	my @params = @_;
	if ($#params != 1) {
		print STDERR "set_sqare_pos: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%.2f %.2f", $params[0], $params[1]);
	my ($resp,$cmd) = lg_chat($command_val{SET_GUIDER_OVLS_POS}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub set_reticle_pos {
	my @params = @_;
	if ($#params != 1) {
		print STDERR "set_sqare_pos: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%.2f %.2f", $params[0], $params[1]);
	my ($resp,$cmd) = lg_chat($command_val{SET_GUIDER_RETICLE_POS}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


sub find_star {
	my @params = @_;
	if ($#params >= 0) {
		print STDERR "find_star: Wrong parameters.\n";
		return undef;
	}
	my ($resp,$cmd) = lg_chat($command_val{FIND_STAR} ,"");
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^Error/) {
		return undef;
	} else {
		return 1;
	}
}


sub save_frame {
	my @params = @_;
	if ($#params != 0) {
		print STDERR "save_frame: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%s", $params[0]);
	my ($resp,$cmd) = lg_chat($command_val{SAVE_FRAME}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^SAVED/) {
		return 1;
	} else {
		return undef;
	}
}

sub save_frame_decorated {
	my @params = @_;
	if ($#params != 0) {
		print STDERR "save_frame_decorated: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%s", $params[0]);
	my ($resp,$cmd) = lg_chat($command_val{SAVE_FRAME_DECORATED}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^SAVED/) {
		return 1;
	} else {
		return undef;
	}
}


sub set_dithering_range {
	my @params = @_;
	if ($#params != 0) {
		print STDERR "set_dithering_range: Wrong parameters.\n";
		return undef;
	}
	my $paramstr = sprintf("%.2f", $params[0]);
	my ($resp,$cmd) = lg_chat($command_val{SET_DITHERING_RANGE}, $paramstr);
	print "$command_name{$cmd} -> $resp\n";
	if ($resp =~ /^OK/) {
		return 1;
	} else {
		return undef;
	}
}


# main routine
sub main {
	my %options = ();

	my $command = shift @ARGV;

	$sock_name = SOCK_NAME;
	if (defined $ENV{LG_SOCKET}) {
		$sock_name = $ENV{LG_SOCKET};
	}

	if (getopts("vh", \%options) == undef) {
		exit 2;
	}

	if (defined($options{h}) or (!defined($command))) {
		print_help();
		exit 0;
	}

	if(defined($options{v})) {
		$verbose = 1;
	}

	if ($command eq "-v") {  # allow -v before cmmand
		$verbose = 1;
		$command = shift @ARGV;
	}

	if ($command eq "get_ver") {
		if (! get_ver(@ARGV)) {
			$verbose && print STDERR "Get info returned error.\n";
			exit 1;
		}
		$verbose && print "Get info succeeded.\n";
		exit 0;

	} elsif ($command eq "guider") {
		if (! guider(@ARGV)) {
			$verbose && print STDERR "GUIDER returned error.\n";
			exit 1;
		}
		$verbose && print "GUIDER succeeded.\n";
		exit 0;

	} elsif ($command eq "status") {
		if (! status(@ARGV)) {
			$verbose && print STDERR "GET_GUIDER_STATE returned error.\n";
			exit 1;
		}
		$verbose && print "GET_GUIDER_STATE succeeded.\n";
		exit 0;

	} elsif ($command eq "set_dithering_range") {
		if (! set_dithering_range(@ARGV)) {
			$verbose && print STDERR "SET_DITHERING_RANGE returned error.\n";
			exit 1;
		}
		$verbose && print "FIND_STAR succeeded.\n";
		exit 0;

	} elsif ($command eq "dither") {
		if (! dither(@ARGV)) {
			$verbose && print STDERR "DITHER returned error.\n";
			exit 1;
		}
		$verbose && print "DITHER succeeded.\n";
		exit 0;

	}  elsif ($command eq "dither_no_wait") {
		if (! dither_no_wait(@ARGV)) {
			$verbose && print STDERR "DITHER_NO_WAIT returned error.\n";
			exit 1;
		}
		$verbose && print "DITHER_NO_WAIT succeeded.\n";
		exit 0;

	} elsif ($command eq "get_distance") {
		if (! get_distance(@ARGV)) {
			$verbose && print STDERR "GET_DISTANCE returned error.\n";
			exit 1;
		}
		$verbose && print "GET_DISTANCE succeeded.\n";
		exit 0;

	} elsif ($command eq "get_ra_dec_drift") {
		if (! get_ra_dec_drift(@ARGV)) {
			$verbose && print STDERR "GET_RA_DEC_DRIFT returned error.\n";
			exit 1;
		}
		$verbose && print "GET_RA_DEC_DRIFT succeeded.\n";
		exit 0;

	} elsif ($command eq "set_square_pos") {
		if (! set_square_pos(@ARGV)) {
			$verbose && print STDERR "SET_GUIDER_SQUARE_POS returned error.\n";
			exit 1;
		}
		$verbose && print "SET_GUIDER_SQUARE_POS succeeded.\n";
		exit 0;

	} elsif ($command eq "set_ovls_pos") {
		if (! set_square_pos(@ARGV)) {
			$verbose && print STDERR "SET_VISIBLE_OVLS_POS returned error.\n";
			exit 1;
		}
		$verbose && print "SET_GUIDER_SQUARE_POS succeeded.\n";
		exit 0;

	} elsif ($command eq "set_reticle_pos") {
		if (! set_reticle_pos(@ARGV)) {
			$verbose && print STDERR "SET_RETICLE_POS returned error.\n";
			exit 1;
		}
		$verbose && print "SET_GUIDER_SQUARE_POS succeeded.\n";
		exit 0;

	} elsif ($command eq "find_star") {
		if (! find_star(@ARGV)) {
			$verbose && print STDERR "FIND_STAR returned error.\n";
			exit 1;
		}
		$verbose && print "FIND_STAR succeeded.\n";
		exit 0;

	} elsif ($command eq "save_frame") {
		if (! save_frame(@ARGV)) {
			$verbose && print STDERR "SAVE_FRAME returned error.\n";
			exit 1;
		}
		$verbose && print "SAVE_FRAME succeeded.\n";
		exit 0;

	} elsif ($command eq "save_frame_decorated") {
		if (! save_frame_decorated(@ARGV)) {
			$verbose && print STDERR "SAVE_FRAME_DECORATED returned error.\n";
			exit 1;
		}
		$verbose && print "SAVE_FRAME_DECRATED succeeded.\n";
		exit 0;

	} elsif ($command eq "-h") {
		print_help();
		exit 0;

	} else {
		print STDERR "There is no such command \"$command\".\n";
		exit 3;
	}
} main;
